(DOCSP-29261): Authentication (#80)
mongodb/docs-kotlin: source/fundamentals/auth.txt

* :ref:`MONGODB-AWS <mongodb-aws-auth-mechanism>`
* :ref:`X.509 <x509-auth-mechanism>`

.. TODO (DOCSP-29270) Add back when Enterprise Auth is available
.. To authenticate using ``Kerberos`` or ``LDAP``, see the
.. :doc:`Enterprise Authentication Mechanisms guide </fundamentals/enterprise-auth>`.

For more information on establishing a connection to your MongoDB cluster,
read our :doc:`Connection Guide </fundamentals/connection>`.

> 

Reviewer: I can't comment on it, but the Rst included in line 45 links to the Java api documentation instead of the kotlin coroutine docs

Author: Good catch. Updated

      string, omit the mechanism. Your code to instantiate a ``MongoClient``
      should look something like this:

      .. literalinclude:: /examples/generated/AuthTest.snippet.default-cred-string.kt
         :language: kotlin
      
   .. tab::
      :tabid: MongoCredential

      To specify the default authentication mechanism using the
      ``MongoCredential`` class, use the ``createCredential()`` method. Your

>       code to instantiate a ``MongoClient`` should look something like this:

Reviewer: ```suggestion
      code to instantiate a ``MongoClient`` should resemble the following:
```

Author: Updated throughout

      .. literalinclude:: /examples/generated/AuthTest.snippet.default-mongo-cred.kt
         :language: kotlin

For more information on the challenge-response (CR) and salted

> challenge-response authentication mechanisms (SCRAM) that MongoDB supports,

Reviewer: Can't comment, but in the line below change server manual to Server manual

Author: Updated

> AWS SDK for Java

Author: Updated

#. Supply your credentials using one of the methods in the credential
   provider chain

To specify the authentication mechanism by using a ``MongoCredential``,
use the ``MongoCredential.createAwsCredential()`` factory method
and add the ``MongoCredential`` instance to your ``MongoClient`` as shown
in the following example:

.. literalinclude:: /examples/generated/AuthTest.snippet.aws-credential.kt
   :language: kotlin
   :emphasize-lines: 1, 9

To specify the authentication mechanism in the connection string, add

> it as a parameter as shown in the following example:

Reviewer: Question: why does this sub section not include the two sample auth code snippets in tabs like elsewhere on the page?

Author: Unsatisfying answer: we were avoiding reformatting/rewriting unless there was an actual error, and this is the format in the Java doc

and add the ``MongoCredential`` instance to your ``MongoClient`` as shown
in the following example:

.. literalinclude:: /examples/generated/AuthTest.snippet.aws-credential-mongo-cred.kt
   :language: kotlin

>    :emphasize-lines: 1

Reviewer: Also emphasize line 9 (`.credential(credential)`)

Author: Updated

and add the ``MongoCredential`` instance to your ``MongoClient`` as shown
in the following example:

.. literalinclude:: /examples/generated/AuthTest.snippet.aws-credential-mongo-cred.kt
   :language: kotlin
   :emphasize-lines: 1

If you need to specify an AWS session token, pass it to the
`withMechanismProperty() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/MongoCredential.html#withMechanismProperty(java.lang.String,T)>`__
method as shown in the following example:

.. literalinclude:: /examples/generated/AuthTest.snippet.aws-credential-session-token.kt
   :language: kotlin
   :emphasize-lines: 1, 2, 10

To refresh your credentials, you can declare a ``Supplier`` lambda expression
that returns new credentials as shown in the following example:

.. literalinclude:: /examples/generated/AuthTest.snippet.aws-refresh-credentials.kt
   :language: kotlin
   :emphasize-lines: 4-6, 9

> .. note::

Reviewer: Suggestion: move this out of the note into reg text

Author: Done
